File: analytics/src/test/java/io/appmetrica/analytics/impl/ExtraMetaInfoRetrieverTest.java
package io.appmetrica.analytics.impl;

import io.appmetrica.analytics.testutils.CommonTest;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ExtraMetaInfoRetrieverTest extends CommonTest {

    @Mock
    private StringResourceRetriever mStringResourceRetriever;
    @Mock
    private BooleanResourceRetriever mBooleanResourceRetriever;
    private ExtraMetaInfoRetriever mExtraMetaInfoRetriever;

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        mExtraMetaInfoRetriever = new ExtraMetaInfoRetriever(mStringResourceRetriever, mBooleanResourceRetriever);
    }

    @Test
    public void testNoResources() {
        when(mStringResourceRetriever.getResource()).thenReturn(null);
        when(mBooleanResourceRetriever.getResource()).thenReturn(null);
        assertThat(mExtraMetaInfoRetriever.getBuildId()).isNull();
        assertThat(mExtraMetaInfoRetriever.isOffline()).isNull();
    }

    @Test
    public void testHasResources() {
        final String buildId = "1234567890";
        final boolean isOffline = false;
        when(mStringResourceRetriever.getResource()).thenReturn(buildId);
        when(mBooleanResourceRetriever.getResource()).thenReturn(isOffline);
        assertThat(mExtraMetaInfoRetriever.getBuildId()).isEqualTo(buildId);
        assertThat(mExtraMetaInfoRetriever.isOffline()).isEqualTo(isOffline);
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/ExtraMetaInfoRetrieverTest.kt
package io.appmetrica.analytics.impl

import android.content.Context
import android.os.Bundle
import io.appmetrica.analytics.coreutils.internal.services.SafePackageManager
import io.appmetrica.analytics.testutils.CommonTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

internal class ExtraMetaInfoRetrieverTest : CommonTest() {

    private val bundle: Bundle = mock()
    private val context: Context = mock()
    private val packageManager: SafePackageManager = mock {
        on { getApplicationMetaData(context) } doReturn bundle
    }
    private val stringResourceRetriever: StringResourceRetriever = mock()
    private val booleanResourceRetriever: BooleanResourceRetriever = mock()

    private val pluginIdKey = "io.appmetrica.analytics.plugin_id"
    private val pluginAdRevenueMetaInfoSourcesApiKey = "io.appmetrica.analytics.plugin_supported_ad_revenue_sources"

    private val extraMetaInfoRetriever: ExtraMetaInfoRetriever by setUp {
        ExtraMetaInfoRetriever(context, stringResourceRetriever, booleanResourceRetriever, packageManager)
    }

    @Test
    fun noResources() {
        whenever(stringResourceRetriever.resource).thenReturn(null)
        whenever(booleanResourceRetriever.resource).thenReturn(null)
        assertThat(extraMetaInfoRetriever.buildId).isNull()
        assertThat(extraMetaInfoRetriever.isOffline).isNull()
    }

    @Test
    fun hasResources() {
        val buildId = "1234567890"
        val isOffline = false
        whenever(stringResourceRetriever.resource).thenReturn(buildId)
        whenever(booleanResourceRetriever.resource).thenReturn(isOffline)
        assertThat(extraMetaInfoRetriever.buildId).isEqualTo(buildId)
        assertThat(extraMetaInfoRetriever.isOffline).isEqualTo(isOffline)
    }

    @Test
    fun `pluginId for null meta data`() {
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(null)
        assertThat(extraMetaInfoRetriever.pluginId).isNull()
    }

    @Test
    fun `pluginId for empty meta data`() {
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(Bundle())
        assertThat(extraMetaInfoRetriever.pluginId).isNull()
    }

    @Test
    fun `pluginId for empty value`() {
        whenever(bundle.getString(pluginIdKey)).thenReturn("")
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(bundle)
        assertThat(extraMetaInfoRetriever.pluginId).isEmpty()
    }

    @Test
    fun `pluginId for valid value`() {
        val pluginId = "Some plugin id"
        whenever(bundle.getString(pluginIdKey)).thenReturn(pluginId)
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(bundle)
        assertThat(extraMetaInfoRetriever.pluginId).isEqualTo(pluginId)
    }

    @Test
    fun `pluginAdRevenueMetaInfoSources for null meta data`() {
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(null)
        assertThat(extraMetaInfoRetriever.pluginAdRevenueMetaInfoSources).isNull()
    }

    @Test
    fun `pluginAdRevenueMetaInfoSources for empty meta data`() {
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(Bundle())
        assertThat(extraMetaInfoRetriever.pluginAdRevenueMetaInfoSources).isNull()
    }

    @Test
    fun `pluginAdRevenueMetaInfoSources for empty value`() {
        whenever(bundle.getString(pluginAdRevenueMetaInfoSourcesApiKey)).thenReturn("")
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(bundle)
    }

    @Test
    fun `pluginAdRevenueMetaInfoSources for valid value`() {
        val pluginAdRevenueMetaInfoSources = "Some plugin ad revenue meta info sources"
        whenever(bundle.getString(pluginAdRevenueMetaInfoSourcesApiKey)).thenReturn(pluginAdRevenueMetaInfoSources)
        whenever(packageManager.getApplicationMetaData(context)).thenReturn(bundle)
        assertThat(extraMetaInfoRetriever.pluginAdRevenueMetaInfoSources).isEqualTo(pluginAdRevenueMetaInfoSources)
    }
}
